File: backend/src/services/places/placesService.ts
import axios from 'axios';

const OPENTRIPMAP_BASE = 'https://api.opentripmap.com/0.1/en/places';
const GEODB_BASE = 'https://wft-geo-db.p.rapidapi.com/v1/geo';

export interface PlaceResult {
  id: string;
  name: string;
  location: string;
  country: string;
  state: string;
  description: string;
  type: string;
  coordinates: { lat: number; lon: number };
  imageUrl: string;
  searchTerms: string[];
  rating: number;
  kinds: string;
}

export interface AutocompleteResult {
  id: string;
  name: string;
  location: string;
  country: string;
  state: string;
  description: string;
  type: string;
  coordinates: { lat: number; lon: number };
  searchTerms: string[];
  population?: number;
}

function getTypeLabel(type: string): string {
  switch (type?.toLowerCase()) {
    case 'major_city': return 'Major City';
    case 'city': return 'City';
    case 'town': return 'Town';
    case 'capital': return 'Capital City';
    default: return 'Place';
  }
}

export class PlacesService {
  private get otmApiKey(): string {
    return process.env.OPENTRIPMAP_API_KEY || '';
  }
  private get geodbApiKey(): string {
    return process.env.GEODB_API_KEY || '';
  }
  private get geodbHost(): string {
    return process.env.GEODB_HOST || '';
  }

  async searchPlaces(query: string, limit: number = 10): Promise<PlaceResult[]> {
    if (!this.otmApiKey) return [];

    const queryVariations = [query, `${query} city`, query.replace(/\s+/g, '')];

    let lat = 0, lon = 0, locationName = '', country = '', state = '';

    for (const qv of queryVariations) {
      try {
        const geoRes = await axios.get(`${OPENTRIPMAP_BASE}/geoname`, {
          params: { name: qv, apikey: this.otmApiKey },
          timeout: 5000,
        });
        if (geoRes.data?.lat) {
          ({ lat, lon, name: locationName, country } = geoRes.data);
          state = geoRes.data.state || '';
          break;
        }
      } catch { continue; }
    }

    if (!lat || !lon) return [];

    await new Promise((r) => setTimeout(r, 100));

    const placesRes = await axios.get(`${OPENTRIPMAP_BASE}/radius`, {
      params: {
        radius: 10000, lon, lat,
        limit: Math.min(limit, 20),
        format: 'geojson', rate: 3,
        apikey: this.otmApiKey,
      },
      timeout: 10000,
    });

    const features = placesRes.data?.features;
    if (!features || !Array.isArray(features)) return [];

    const places: PlaceResult[] = [];
    for (let i = 0; i < Math.min(features.length, limit); i++) {
      const feature = features[i];
      const props = feature.properties;
      const coords = feature.geometry.coordinates;

      if (i > 0) await new Promise((r) => setTimeout(r, 200));

      let details: any = null;
      try {
        if (props.xid) {
          const detailRes = await axios.get(`${OPENTRIPMAP_BASE}/xid/${props.xid}`, {
            params: { apikey: this.otmApiKey },
            timeout: 5000,
          });
          details = detailRes.data;
        }
      } catch { /* skip */ }

      const kinds = props.kinds || '';
      let type = 'attraction';
      if (kinds.includes('museums')) type = 'museum';
      else if (kinds.includes('historic')) type = 'historic';
      else if (kinds.includes('natural')) type = 'natural';
      else if (kinds.includes('cultural')) type = 'cultural';
      else if (kinds.includes('religion') || kinds.includes('architecture')) type = 'historic';

      places.push({
        id: props.xid || `otm_${Math.random().toString(36).slice(2, 11)}`,
        name: props.name || 'Unknown Place',
        location: locationName,
        country: country || '',
        state,
        description: details?.wikipedia_extracts?.text?.slice(0, 200) ||
          details?.info?.descr || kinds.split(',').join(', ') || 'Tourist attraction',
        type,
        coordinates: { lat: coords[1], lon: coords[0] },
        imageUrl: details?.preview?.source || details?.image ||
          `https://source.unsplash.com/800x600/?${encodeURIComponent(props.name + ' ' + locationName)}`,
        searchTerms: [
          props.name?.toLowerCase(),
          locationName?.toLowerCase(),
          country?.toLowerCase(),
          ...kinds.split(',').filter((k: string) => k.trim()),
        ].filter(Boolean) as string[],
        rating: props.rate || 0,
        kinds,
      });
    }

    return places
      .filter((p) => p.name && p.name !== 'Unknown Place')
      .sort((a, b) => (b.rating || 0) - (a.rating || 0));
  }

  async getAutocomplete(query: string, limit: number = 8): Promise<AutocompleteResult[]> {
    if (!this.geodbApiKey || !query || query.length < 2) return [];

    try {
      const res = await axios.get(`${GEODB_BASE}/cities`, {
        params: {
          namePrefix: query, limit,
          minPopulation: 1000, sort: '-population',
          languageCode: 'en',
        },
        headers: {
          'X-RapidAPI-Key': this.geodbApiKey,
          'X-RapidAPI-Host': this.geodbHost,
        },
        timeout: 5000,
      });

      if (!res.data?.data) return [];

      return res.data.data.map((city: any) => {
        let type = 'city';
        if (city.population > 1000000) type = 'major_city';
        else if (city.population > 100000) type = 'city';
        else if (city.population > 10000) type = 'town';
        else type = 'place';

        const locationParts = [city.region, city.country].filter(Boolean);
        const locationDesc = locationParts.join(', ');

        return {
          id: `geodb_${city.id}`,
          name: city.name,
          location: locationDesc,
          country: city.country || '',
          state: city.region || '',
          description: `${getTypeLabel(type)}${locationDesc ? ` in ${locationDesc}` : ''}${city.population ? ` (pop. ${city.population.toLocaleString()})` : ''}`,
          type,
          coordinates: { lat: city.latitude, lon: city.longitude },
          searchTerms: [city.name?.toLowerCase(), city.country?.toLowerCase(), city.region?.toLowerCase()].filter(Boolean) as string[],
          population: city.population,
        };
      });
    } catch {
      return [];
    }
  }
}

export const placesService = new PlacesService();

File: backend-python/app/services/places_service.py
"""Places service — OpenTripMap search + GeoDB autocomplete."""

import asyncio
import httpx

from app.config import settings


class PlacesService:
    OTM_BASE = "https://api.opentripmap.com/0.1/en/places"
    GEODB_BASE = "https://wft-geo-db.p.rapidapi.com/v1/geo"

    def _otm_key(self) -> str:
        return settings.opentripmap_api_key

    def _geodb_key(self) -> str:
        return settings.geodb_api_key

    def _geodb_host(self) -> str:
        return settings.geodb_host

    async def search_places(self, query: str, limit: int = 10) -> list[dict]:
        if not self._otm_key():
            return []

        query_variations = [query, f"{query} city", query.replace(" ", "")]
        lat, lon, location_name, country, state = 0, 0, "", "", ""

        async with httpx.AsyncClient(timeout=10.0) as client:
            for qv in query_variations:
                try:
                    resp = await client.get(
                        f"{self.OTM_BASE}/geoname",
                        params={"name": qv, "apikey": self._otm_key()},
                    )
                    data = resp.json()
                    if data.get("lat"):
                        lat = data["lat"]
                        lon = data["lon"]
                        location_name = data.get("name", "")
                        country = data.get("country", "")
                        state = data.get("state", "")
                        break
                except Exception:
                    continue

            if not lat or not lon:
                return []

            await asyncio.sleep(0.1)

            try:
                resp = await client.get(
                    f"{self.OTM_BASE}/radius",
                    params={
                        "radius": 10000, "lon": lon, "lat": lat,
                        "limit": min(limit, 20),
                        "format": "geojson", "rate": 3,
                        "apikey": self._otm_key(),
                    },
                )
                places_data = resp.json()
            except Exception:
                return []

            features = places_data.get("features", [])
            if not features:
                return []

            places = []
            for i, feature in enumerate(features[:limit]):
                props = feature.get("properties", {})
                coords = feature.get("geometry", {}).get("coordinates", [0, 0])

                if i > 0:
                    await asyncio.sleep(0.2)

                details = None
                try:
                    if props.get("xid"):
                        detail_resp = await client.get(
                            f"{self.OTM_BASE}/xid/{props['xid']}",
                            params={"apikey": self._otm_key()},
                        )
                        details = detail_resp.json()
                except Exception:
                    pass

                kinds = props.get("kinds", "")
                place_type = "attraction"
                if "museums" in kinds:
                    place_type = "museum"
                elif "historic" in kinds:
                    place_type = "historic"
                elif "natural" in kinds:
                    place_type = "natural"
                elif "cultural" in kinds:
                    place_type = "cultural"
                elif "religion" in kinds or "architecture" in kinds:
                    place_type = "historic"

                wiki_text = ""
                if details:
                    wiki_text = (details.get("wikipedia_extracts", {}) or {}).get("text", "")[:200]

                places.append({
                    "id": props.get("xid", f"otm_{i}"),
                    "name": props.get("name", "Unknown Place"),
                    "location": location_name,
                    "country": country,
                    "state": state,
                    "description": wiki_text or (details or {}).get("info", {}).get("descr", "") or ", ".join(kinds.split(",")) or "Tourist attraction",
                    "type": place_type,
                    "coordinates": {"lat": coords[1], "lon": coords[0]},
                    "imageUrl": (details or {}).get("preview", {}).get("source") or (details or {}).get("image") or f"https://source.unsplash.com/800x600/?{props.get('name', '')}+{location_name}",
                    "searchTerms": [p for p in [props.get("name", "").lower(), location_name.lower(), country.lower()] if p] + [k.strip() for k in kinds.split(",") if k.strip()],
                    "rating": props.get("rate", 0),
                    "kinds": kinds,
                })

            return [p for p in places if p["name"] and p["name"] != "Unknown Place"]

    async def get_autocomplete(self, query: str, limit: int = 8) -> list[dict]:
        if not self._geodb_key() or not query or len(query) < 2:
            return []

        try:
            async with httpx.AsyncClient(timeout=5.0) as client:
                resp = await client.get(
                    f"{self.GEODB_BASE}/cities",
                    params={
                        "namePrefix": query, "limit": limit,
                        "minPopulation": 1000, "sort": "-population",
                        "languageCode": "en",
                    },
                    headers={
                        "X-RapidAPI-Key": self._geodb_key(),
                        "X-RapidAPI-Host": self._geodb_host(),
                    },
                )
                data = resp.json()
                cities = data.get("data", [])
                if not cities:
                    return []

                results = []
                for city in cities:
                    pop = city.get("population", 0)
                    city_type = "major_city" if pop > 1_000_000 else "city" if pop > 100_000 else "town" if pop > 10_000 else "place"
                    location_parts = [p for p in [city.get("region"), city.get("country")] if p]
                    location_desc = ", ".join(location_parts)

                    results.append({
                        "id": f"geodb_{city.get('id')}",
                        "name": city.get("name", ""),
                        "location": location_desc,
                        "country": city.get("country", ""),
                        "state": city.get("region", ""),
                        "description": f"{city_type.replace('_', ' ').title()}{' in ' + location_desc if location_desc else ''}{' (pop. ' + format(pop, ',') + ')' if pop else ''}",
                        "type": city_type,
                        "coordinates": {"lat": city.get("latitude", 0), "lon": city.get("longitude", 0)},
                        "searchTerms": [p for p in [city.get("name", "").lower(), city.get("country", "").lower(), city.get("region", "").lower()] if p],
                        "population": pop,
                    })
                return results
        except Exception:
            return []


places_service = PlacesService()
